Python/Product/Analysis/LanguageServer/Server.cs

```diff
@@ -87,13 +95,47 @@ public sealed class Server : ServerBase, IDisposable {
         #region Client message handling
 
         public override async Task<InitializeResult> Initialize(InitializeParams @params) {
-            _testEnvironment = @params.initializationOptions.interpreter.properties.ContainsKey("TestEnvironment");
-            if (_testEnvironment) {
-                // Test environment needs predictable initialization.
-                // Tests can only proceed when analysis is fully done.
-                _analyzer = CreateAnalyzer(@params.initializationOptions.interpreter).Result;
+            _testEnvironment = @params.initializationOptions.testEnvironment;
+            // Test environment needs predictable initialization.
+            if (@params.initializationOptions.asyncStartup && !_testEnvironment) {
```

Comment:
Do we need `testEnvironment` if we already have `asyncStartup`?

Reply:
There are effectively 3 modes
1. Test mode, all sync
2. VS mode, as it used to work since I am not 100% confident VS part can handle all-async as I don't know all the entry points.
3. VSC mode to minimize delays. VSC only need synchronization at the LS entry points. This is what asyncStartup controls.

VSC passes true for asyncStartup  and true/false for tests. asyncStartup == false does not make it completely sync, VS still uses awaits in places. VSC creates analyzer async while VS does not.

if VS only uses LS entry points then we can just do (3). But bunch of methods are internal with visibility so I don't know.